modules/apps/transfer/keeper/relay.go

```diff
@@ -214,72 +226,115 @@ func (k Keeper) OnRecvPacket(ctx sdk.Context, packet channeltypes.Packet, data t
-			ctx, types.ModuleName, sdk.NewCoins(voucher),
-		); err != nil {
-			return errorsmod.Wrap(err, "failed to mint IBC tokens")
-		}
+			// send to receiver
+			if k.IsBlockedAddr(receiver) {
+				return errorsmod.Wrapf(ibcerrors.ErrUnauthorized, "%s is not allowed to receive funds", receiver)
+			}
+			moduleAddr := k.authKeeper.GetModuleAddress(types.ModuleName)
+			if moduleAddr == nil {
+				return errorsmod.Wrapf(sdkerrors.ErrUnknownAddress, "module account %s does not exist", types.ModuleName)
+			}
+			if err := k.bankKeeper.SendCoins(
+				ctx, moduleAddr, receiver, sdk.NewCoins(voucher),
+			); err != nil {
+				return errorsmod.Wrapf(err, "failed to send coins to receiver %s", receiver.String())
+			}
 
-		// send to receiver
-		if err := k.bankKeeper.SendCoinsFromModuleToAccount(
-			ctx, types.ModuleName, receiver, sdk.NewCoins(voucher),
-		); err != nil {
-			return errorsmod.Wrapf(err, "failed to send coins to receiver %s", receiver.String())
+			denomPath := token.Denom.Path()
+			labels = append(labels, telemetry.NewLabel(coretypes.LabelSource, "false"))
+			defer internaltelemetry.ReportOnRecvPacketTelemetry(transferAmount, denomPath, labels)
+
+			receivedCoins = append(receivedCoins, voucher)
 		}
+	}
 
-		denomPath := token.Denom.Path()
-		labels = append(labels, telemetry.NewLabel(coretypes.LabelSource, "false"))
-		defer internaltelemetry.ReportOnRecvPacketTelemetry(transferAmount, denomPath, labels)
+	if data.ShouldBeForwarded() {
+		// we are now sending from the forward escrow address to the final receiver address.
+		if err := k.forwardPacket(ctx, data, packet, receivedCoins); err != nil {
+			return err
+		}
 	}
 
+	// The ibc_module.go module will return the proper ack.
 	return nil
 }
 
-// OnAcknowledgementPacket responds to the success or failure of a packet
-// acknowledgement written on the receiving chain. If the acknowledgement
-// was a success then nothing occurs. If the acknowledgement failed, then
-// the sender is refunded their tokens using the refundPacketTokens function.
+// OnAcknowledgementPacket either reverts the state changes executed in receive
+// and send packet if the chain acted as a middle hop on a multihop transfer; or
+// responds to the success or failure of a packet acknowledgement written on the
+// final receiving chain, if it acted as the original sender chain. If the
+// acknowledgement was a success then nothing occurs. If the acknowledgement failed,
+// then the sender is refunded their tokens using the refundPacketToken function.
 func (k Keeper) OnAcknowledgementPacket(ctx sdk.Context, packet channeltypes.Packet, data types.FungibleTokenPacketDataV2, ack channeltypes.Acknowledgement) error {
+	prevPacket, isForwarded := k.GetForwardedPacket(ctx, packet.SourcePort, packet.SourceChannel, packet.Sequence)
+
 	switch ack.Response.(type) {
 	case *channeltypes.Acknowledgement_Result:
+		if isForwarded {
+			return k.ackForwardPacketSuccess(ctx, prevPacket, packet)
```

Comment:
this doesn't need to be a helper fn

Reply:
I agree with the general sentiment. I also find there's an excessive use of smaller funcs too

Reply:
For context: the pros and cons of having these helper functions were discussed [here](https://github.com/cosmos/ibc-go/pull/6574#discussion_r1636013788).